### PR TITLE
Fixes #25888 - move components from Katello to Foreman

### PR DESCRIPTION
**Pull request by MariaAga**
Add the ~~'ConfirmDialog', 'Dialog',~~ 'ModalProgressBar' and 'Settings' Components to foreman
Add some of the helper functions to foreman's helper

Because Katello is no longer using  'ConfirmDialog', 'Dialog' ([PR](https://github.com/Katello/katello/pull/7970)) there is no reason to move it to foreman, so it will not be moved.

### CODE REVIEW

File: webpack/assets/javascripts/bundle.js [hunk 1]
```diff
@@ -40,7 +40,7 @@ window.tfm = Object.assign(window.tfm || {}, {
   hostgroups: require('./foreman_hostgroups'),
   hosts: require('./foreman_hosts'),
   httpProxies: require('./foreman_http_proxies'),
-  toastNotifications: require('./foreman_toast_notifications'),
+  toastNotifications: require('./react_app/common/foreman_toast_notifications'),
```
Comment (ohadlevy) [inline webpack/assets/javascripts/bundle.js]: Why? this is used outside of react.

File: webpack/assets/javascripts/react_app/common/helpers.js [hunk 2]
```diff
@@ -51,6 +51,24 @@ export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
 export const getDisplayName = Component =>
   Component.displayName || Component.name || 'Component';
 
+export const urlBuilder = (controller, action, id = undefined) =>
+  `/${controller}/${id ? `${id}/` : ''}${action}`;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/helpers.js]: why is this required? this is missing a test. maybe we should create a url helper file instead of putting it all under helpers.

File: webpack/assets/javascripts/react_app/common/helpers.js [hunk 3]
```diff
@@ -51,6 +51,24 @@ export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
...
+
+export const urlWithSearch = (base, searchQuery) =>
+  `/${base}?search=${searchQuery}`;
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/helpers.js]: is this is required? /cc @Laviro

File: webpack/assets/javascripts/react_app/common/helpers.js [hunk 4]
```diff
@@ -51,6 +51,24 @@ export const clearSpaces = string => string.trim().replace(/\s\s+/, ' ');
...
+export const stringIsInteger = value => {
+  // checking for positive integers only
+  const reg = new RegExp('^[0-9]+$');
+  return reg.test(value);
+};
+
+export const KEY_CODES = {
+  TAB_KEY: 9,
+  ENTER_KEY: 13,
+  ESCAPE_KEY: 27,
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/common/helpers.js]: @Laviro we already have it defined somewhere right?

File: webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js [hunk 5]
```diff
@@ -0,0 +1,43 @@
+import axios from 'axios';
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js]: is that required? /cc @amirfefer ?

File: webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Button } from 'patternfly-react';
+import { translate as __ } from '../../../common/I18n';
+import Dialog from '../Dialog';
+
+const ConfirmDialog = props => {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js]: @boaz0 does that overlap with your delete dialog in #6301?

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+import PropTypes from 'prop-types';
+import { Modal, ProgressBar } from 'patternfly-react';
+import { sprintf, translate as __ } from '../../../common/I18n';
+
+const ModalProgressBar = props => {
+  const { show, container, task } = props;
+  let modalTitle = null;
+  let progress = 0;
+
+  if (task) {
```
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js]: i wonder if this should be in tasks plugin, or perhaps a different component wrapping this one?

**Review by ohadlevy — DISMISSED** [review]
thanks @MariaAga for your contribution.

there are many components in here, IMHO it would be better to create a PR for each, regardless, I've tried to review it. I would be happy if we would actually use it in foreman code vs just adding components. regardless, they should appear in our storybook.

I would like to ask @theforeman/ui-ux also to review the code.

File: webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js (hunk 6 above)
Comment (boaz0) [reply webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js]: It seems like it doesn't really overlap. However, I am not sure about reusing this.
@MariaAga why did you choose `Modal` over [`MessageDialog`](http://patternfly-react.surge.sh/patternfly-3/index.html?knob-Icon=info&knob-With%20Title=true&knob-With%20Primary%20Content=true&knob-With%20Secondary%20Button=true&knob-With%20Icon=true&knob-With%20Secondary%20Content=true&selectedKind=patternfly-react%2FCommunication%2FMessage%20Dialog&selectedStory=Message%20Dialog%20with%20Toggleable%20Options&full=0&addons=1&stories=1&panelRight=0&addonPanel=storybooks%2Fstorybook-addon-knobs)?

Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js]: @boaz0 if you mean using `Modal` in the `Dialog` component then I didn't choose `Modal`, it was in the code, and in this PR I only moved components from Katello to Foreman without changing them.

Comment (boaz0) [reply webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js]: @MariaAga I see. That's what I thought.
Anyway, to me it looks like that PF3 has implemented this component: `MessageDialog` and we should use it instead of this.

If we can do this in here then great. Otherwise, lets see how we can align with PF3 as soon as possible.

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 4 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/common/helpers.js]: We have it inside the `AutoCompleteConstants` component, but we can use this helpers now.

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 3 above)
Comment (Ron-Lavi) [reply webpack/assets/javascripts/react_app/common/helpers.js]: In the `AutoComplete` I've used the `URI` library,
for example:
```
const APIPath = new URI(url);
  APIPath.addSearch({ search: APISearchQuery });
  return APIPath.toString();
```

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js (hunk 7 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js]: @iNecas what do you think?

File: webpack/assets/javascripts/bundle.js (hunk 1 above)
Comment (MariaAga) [reply webpack/assets/javascripts/bundle.js]: cc  @amirfefer

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js (hunk 7 above)
Comment (iNecas) [reply webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js]: We plan introducing react/redux into foreman-tasks, so this code could be moved there once the infra is there. //cc @sharvit

File: webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js (hunk 6 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/ConfirmDialog/ConfirmDialog.js]: Thanks @boaz0, I opened a [separate PR ](https://github.com/Katello/katello/pull/7970) to add `MessageDialog` to Katello

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 4 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/common/helpers.js]: IMHO we should have a `keyCodes.js` file.

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js (hunk 7 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js]: IMHO we can remove lines 11-14 and just receive the `modalTitle` and the `progress` as props.
Those lines should move to Katello so when consuming this component, Katello can pass the 2 relevant props instead of the whole `task` object.

That's way, we will have a more reusable `ModalProgressBar` that not only work with tasks.

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js [hunk 8]
```diff
@@ -0,0 +1,43 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { Modal, ProgressBar } from 'patternfly-react';
+import { sprintf, translate as __ } from '../../../common/I18n';
+
+const ModalProgressBar = props => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.js]:
```suggestion
const ModalProgressBar = ({ show, container, title, progress }) => {
```

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.scss [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+}
+
+.modal-container .modal,
+.modal-container .modal-backdrop {
+  position: absolute;
+}
+
+.modal-container.modal-open {
+  z-index: 0;
+}
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.scss:16]: This file will be much more readable with `scss` syntax:
```scss
#modal-progress-bar {
  min-height: 100px;
}

 // Needed to display modal within a container
.modal-container {
  position: relative;

  .modal, .modal-backdrop {
    position: absolute;
  }

  &.modal-open {
    z-index: 0;
  }
}
```

File: webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+import configureMockStore from 'redux-mock-store';
+import thunk from 'redux-thunk';
+import Immutable from 'seamless-immutable';
+import { mock as mockApi } from '../../../mockRequests';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js]: I would like to suggest a different approach which I see as better, instead of using the `axios` mock, use `jest` mocking to mock the API itself.

See: https://github.com/theforeman/foreman/blob/c18d67464d81943b77feeb0e24ec649574860ab7/webpack/assets/javascripts/react_app/components/BreadcrumbBar/__tests__/BreadcrumbBarActions.test.js

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.scss (hunk 9 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/ModalProgressBar/ModalProgressBar.scss:16]: Fixed, thanks

File: webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js (hunk 10 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js]: Changed the tests, thanks!

File: webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js (hunk 5 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/Settings/__tests__/SettingsActions.test.js]: Changed the tests to use `jest` mocking to mock the API itself instead of mocking axios

File: webpack/assets/javascripts/bundle.js (hunk 1 above)
Comment (sharvit) [reply webpack/assets/javascripts/bundle.js]: any update?

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 4 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/common/helpers.js]: Added keyCodes.js

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 2 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/common/helpers.js]: added a  urlHelpers file

**Comment by sharvit** [comment]
Thanks @MariaAga, can you please fix the snapshot error? https://travis-ci.org/theforeman/foreman/jobs/504665861#L1780-L1781

**Comment by MariaAga** [comment]
@sharvit - fixed, thanks

File: webpack/assets/javascripts/bundle.js (hunk 1 above)
Comment (sharvit) [reply webpack/assets/javascripts/bundle.js]: @amirfefer ?

File: webpack/assets/javascripts/react_app/common/helpers.js [hunk 11]
```diff
@@ -102,6 +102,15 @@ const propsToCase = (casingFn, errorMsg, ob) => {
   }, {});
 };
 
+/**
+ * Check if a string is a positive integer
+ * @param {String} value - the string
+ */
+export const stringIsInteger = value => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/helpers.js]: nitpick: `stringIsPositiveNumber`

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/index.js [hunk 12]
```diff
@@ -0,0 +1,4 @@
+import ModalProgressBar from './ModalProgressBar';
+import './ModalProgressBar.scss';
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/common/ModalProgressBar/index.js]: Could you move the `ModalProgressBar.scss` import to `ModalProgressBar.js`?
Then this file can be:
```js
export { default } from './ModalProgressBar';
```

**Review by sharvit — CHANGES_REQUESTED** [review]

File: webpack/assets/javascripts/react_app/components/common/ModalProgressBar/index.js (hunk 12 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/components/common/ModalProgressBar/index.js]: yes, thanks

File: webpack/assets/javascripts/react_app/common/helpers.js (hunk 11 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/common/helpers.js]: changed, thanks

File: webpack/assets/javascripts/bundle.js (hunk 1 above)
Comment (MariaAga) [reply webpack/assets/javascripts/bundle.js]: Both of us don't remember why I moved it so i'm moving it back. added foremanJS alias to access it from Katello

File: config/webpack.config.js [hunk 13]
```diff
@@ -120,6 +120,9 @@ module.exports = env => {
         foremanReact:
           path.join(__dirname,
              '../webpack/assets/javascripts/react_app'),
+        foremanJS:
```
Comment (sharvit) [inline config/webpack.config.js]: Why is this needed?

File: webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js [hunk 14]
```diff
@@ -0,0 +1,3 @@
+export const notify = () => {};
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js]: Does it actually needed? why?

File: config/webpack.config.js (hunk 13 above)
Comment (MariaAga) [reply config/webpack.config.js]: For Katello, so it can use `foreman_toast_notifications.js `

File: webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js (hunk 14 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js]: No, thanks, its a leftover

File: config/webpack.config.js (hunk 13 above)
Comment (sharvit) [reply config/webpack.config.js]: We are facing an issue with webpack where `foremanReact` get copied into the build results of each plugin. I afraid after this change `foremanJS` will get copied as well.

IMO it is better to use the `window.tfm.toastNotifications` until the issue will get fixed.

Comment (MariaAga) [reply config/webpack.config.js]: Or maybe to leave the small (24 line) `foreman_toast_notifications.js` file in katello?

Comment (sharvit) [reply config/webpack.config.js]: I believe it is better to remove it from katello and use it from the global tfm object so it will be easier to maintain it. Notice plugin developers copy solutions from katello as they are best practices (while they are not). I am afraid to see this code in more plugins.

Comment (MariaAga) [reply config/webpack.config.js]: Sure, removed here and changed in the [katello PR](https://github.com/Katello/katello/pull/7965)

File: webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js (hunk 14 above)
Comment (sharvit) [reply webpack/assets/javascripts/react_app/common/__mocks__/foreman_toast_notifications.js]: So can you remove it please? :smile:

File: config/webpack.config.js [hunk 15]
```diff
@@ -36,7 +36,7 @@ const supportedLocales = () => {
```
Comment (sharvit) [inline config/webpack.config.js]: Does this file change any logic? If it's only style changes I bealive it should be done in a different PR

File: webpack/assets/javascripts/react_app/common/helpers.js [hunk 16]
```diff
@@ -128,6 +128,15 @@ const propsToCase = (casingFn, errorMsg, ob) => {
   }, {});
 };
 
+/**
+ * Check if a string is a positive integer
+ * @param {String} value - the string
+ */
+export const stringIsPositiveNumber = value => {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/helpers.js:135]: Need to be tested

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
+/**
+ * Build a url from given controller, action and id
+ * @param {String} controller - the controller
+ * @param {String} action - the action
+ */
+export const urlBuilder = (controller, action, id = undefined) =>
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:6]: Need to be tested

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+  `/${controller}/${id ? `${id}/` : ''}${action}`;
+
+/**
+ * Build a url with search query
+ * @param {String} base - the base url
+ * @param {String} searchQuery - the search query
+ */
+export const urlWithSearch = (base, searchQuery) =>
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js:14]: Need to be tested

File: webpack/assets/javascripts/react_app/common/urlHelpers.js [hunk 19]
```diff
@@ -0,0 +1,20 @@
...
+  `/${base}?search=${searchQuery}`;
+
+export default {
```
Comment (sharvit) [inline webpack/assets/javascripts/react_app/common/urlHelpers.js]: Do we actually need the `default` export?
Can't we just force consumers to do named imports?

**Review by sharvit — CHANGES_REQUESTED** [review]
Thanks @MariaAga

File: config/webpack.config.js (hunk 15 above)
Comment (MariaAga) [reply config/webpack.config.js]: didn't notice the auto format on save, removing the changes :)

File: webpack/assets/javascripts/react_app/common/urlHelpers.js (hunk 19 above)
Comment (MariaAga) [reply webpack/assets/javascripts/react_app/common/urlHelpers.js]: Removed, thanks

**Comment by sharvit** [comment]
IMO before we can merge this PR, we should make sure https://github.com/Katello/katello/pull/7965 working properly with this change.

**Review by sharvit — DISMISSED** [review]
Thanks @MariaAga LGTM :+1:

**Comment by mmoll** [comment]
@MariaAga could you please squash the commits together, or (even better), split it up to one commit per component? In any case please change the commit message to "Adding x component"

**Comment by MariaAga** [comment]
Squashed the commits, thanks!

**Review by mmoll — APPROVED** [review]

**Comment by mmoll** [comment]
merged, thanks @MariaAga!

**Comment by boaz0** [comment]
Congrats @MariaAga ! You **crashed** it
